test/extensions/quic_listeners/quiche/integration/quic_protocol_integration_test.cc

```diff
@@ -0,0 +1,15 @@
+#include "test/integration/protocol_integration_test.h"
+
+namespace Envoy {
+
+// We do not yet run QUIC downstream tests.
+GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(DownstreamProtocolIntegrationTest);
```

Comment:
Just checking if my mental model works aligns with your intention:
Once I add
```
INSTANTIATE_TEST_SUITE_P(Protocols, ProtocolIntegrationTest,
                         testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParams(
                             {Http::CodecClient::Type::HTTP3},
                             {FakeHttpConnection::Type::HTTP1, Http::CodecClient::Type::HTTP2})),
                         HttpProtocolIntegrationTest::protocolTestParamsToString);
```
The HttpIntegrationTest will try to load a quic specific config and start an envoy quic client connection to run all the tests.

So if I add factory to create EnvoyQuicClientConnection and EnvoyQuicClientSession, then HttpIntegrationTest only need to depends on the factory config rather than EnvoyQuicClientConnection and EnvoyQuicClientSession. So we can compile out QUICHE library in the tests which don't specify Http3.

Is my understanding right?

Reply:
Exactly.  If you instantiate the test, add a client factory, I think you can reuse the downstream tests as well.

We might want to do H1/H2/H3 upstream (but you could TODO the last for the upstream bug) and we'd need a downstream EXCLUDE macro as I'm sure not everything would work off the bat, but it should save you re-implementing a bunch of tests.

Reply:
Thanks for confirming! My plan for downstream test is to make all the tests pass before enabling HTTP3 downstream here. 

A side request, can you move the client factory out of codec? They seem to be not related to each other and it will be easier to use in both cluster manager and downstream integration test.